src/modm/architecture/interface/can_message.hpp

```diff
@@ -76,53 +99,126 @@ struct Message
+
 public:
-	uint32_t identifier;
-	uint8_t modm_aligned(4) data[8];
-	struct Flags
+    uint32_t identifier;
+    struct Flags
 	{
 		Flags() :
-			rtr(0), extended(1)
+			fd(0), brs(0), rtr(0), extended(1)
 		{
 		}
 
+        bool fd : 1;
+        bool brs : 1;
 		bool rtr : 1;
 		bool extended : 1;
 	} flags;
-	uint8_t length;
+	uint8_t dlc;
+	uint8_t modm_aligned(4) data[capacity];
 
 public:
 	inline bool
-	operator == (const modm::can::Message& rhs) const
+	operator == (const modm::can::MessageBase<capacity>& rhs) const
 	{
 		return ((this->identifier     == rhs.identifier) and
-				(this->length         == rhs.length)     and
+				(this->dlc            == rhs.dlc)        and
+                (this->flags.fd       == rhs.flags.fd)   and
+				(this->flags.brs      == rhs.flags.brs)  and
 				(this->flags.rtr      == rhs.flags.rtr)  and
 				(this->flags.extended == rhs.flags.extended) and
-				std::equal(data, data + length, rhs.data));
+				std::equal(data, data + getLength(), rhs.data));
+	}
+
+	inline void
+	operator = (const modm::can::MessageBase<capacity>& rhs)
+	{
+		this->identifier     = rhs.identifier;
+		this->dlc            = rhs.dlc;
+        this->flags.fd       = rhs.flags.fd;
+		this->flags.brs      = rhs.flags.brs;
+		this->flags.rtr      = rhs.flags.rtr;
+		this->flags.extended = rhs.flags.extended;
+		std::copy(std::begin(rhs.data), std::end(rhs.data), std::begin(this->data));
 	}
 
 	inline bool
-	operator < (const modm::can::Message& rhs) const
+	operator < (const modm::can::MessageBase<capacity>& rhs) const
 	{
 		return (this->identifier << (this->flags.extended ? 0 : 18))
 			< (rhs.identifier << (rhs.flags.extended ? 0 : 18));
 	}
+
 };
 
-}	// namespace modm::can
+/// Representation of a CAN message
+/// @ingroup modm_architecture_can
+struct Message : public MessageBase<8>
+{
+	inline Message(uint32_t inIdentifier = 0, uint8_t inLength = 0) :
+		MessageBase(inIdentifier, std::min(inLength, uint8_t(8)))
+	{
+	}
+
+	// Create CAN message from long data in Network Order.
+	inline Message(uint32_t inIdentifier, uint8_t inLength, const uint64_t &inData, bool extended=false) :
+		MessageBase(inIdentifier, std::min(inLength, uint8_t(8)), inData)
+	{
+		flags.extended = extended;
+	}
+
+    inline void
+    setLength(uint8_t inDlc)
+    {
+        dlc = std::min<uint8_t>(inDlc, 8);
+    }
+};
+
+/// Representation of a CAN FD message
+/// @ingroup modm_architecture_can
+struct LongMessage : public modm::can::MessageBase<64>
+{
+    inline LongMessage(uint32_t inIdentifier = 0, uint8_t inDlc = 0) :
+		MessageBase(inIdentifier, std::min(inDlc, uint8_t(15)))
+	{
+	}
+
+	// Create CAN message from long data in Network Order.
+	inline LongMessage(uint32_t inIdentifier, uint8_t inDlc, const uint64_t &inData, bool extended=false) :
+		MessageBase(inIdentifier, std::min(inDlc, uint8_t(15)), inData)
+	{
+        flags.extended = extended;
+	}
+
+    inline void
+    setLength(uint8_t inDlc)
```

Comment:
`setLength(64)` would look fine, but fail silently.

Maybe `setLength()` should take a length as an argument and convert is to the (next bigger) DLC?
And we could add a `setDlc()` function for the user to explicitly set the DLC?

Reply:
I was considering doing this, but I am also very reluctant to introduce more logic than necessary. However, since the code must be backward compatible this might be a better solution.